concurrency/ItemExpiryPurgerTaskMetricsCollector.java

```diff
@@ -0,0 +1,193 @@
+        Integer metricValue = purgerCountMetricsMap.get(metricName);
+        bbAssert(metricValue != null, "addToSingleMetricMapEntry", "Metric not available", "metricName", metricName);
+        purgerCountMetricsMap.put(metricName, metricValue + value);
+    }
+    private synchronized void addToSingleTimeMetricMapEntry (final String metricName, final long value) {
+        ItemExpiryPurgerTimeMetricPair metricValue = purgerTimeMetricsMap.get(metricName);
+        bbAssert(metricValue != null, "addToSingleTimeMetricMapEntry", "Metric not available", "metricName", metricName);
+        metricValue.addItemExpiryPurgerStaleness(value);
+        purgerTimeMetricsMap.put(metricName, metricValue);
+    }
+
+    public void updateItemExpiryPurgerStaleness(final long staleness) {
+        addToSingleTimeMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerStaleness.name(), staleness);
+    }
+
+    public void increasePurgerTaskGetZeroNtpTime() {
+        incrementSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerExpressionZero.name());
+    }
+
+    public void increasePurgerTaskFailedByException() {
+        incrementSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerTaskFailedByException.name());
+    }
+
+    public void increasePurgerTaskThrottledByBaseTableIops() {
+        incrementSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerTaskThrottledByBaseTableIops.name());
+    }
+
+    public void increasePurgerTaskThrottledByGsi() {
+        incrementSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerTaskThrottledByGsi.name());
+    }
+
+    public void increasePurgerTaskThrottledByUpdateStreams() {
+        incrementSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerTaskThrottledByUpdateStreams.name());
+    }
+
+    public void increasePurgerTaskThrottledByReplicaTableEntryThrottle() {
+        incrementSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerTaskThrottledByReplicaTableEntryThrottle.name());
+    }
+
+    public void increasePurgerTaskConditionCheckFailed() {
+        incrementSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerTaskConditionCheckFailed.name());
+    }
+
+    public void increasePurgerTaskUnexpiredItemDequeued() {
+        incrementSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryPurgerTaskUnexpiredItemDequeued.name());
+    }
+
+    public void increaseItemExpiryWriteIOPsPublishedToStreams(int logicalIOCount) {
+        addToSingleCountMetricMapEntry(BBCoralMetrics.Field.ItemExpiryWriteIOPsPublishedToStreams.name(), logicalIOCount /* value */);
+    }
+
+    private void collectCountAndTimeMetrics(final OpContext opContext) {
+        for (Map.Entry<String, Integer> entry : purgerCountMetricsMap.entrySet()) {
+            opContext.getMetrics().addCount(entry.getKey(), entry.getValue(), Unit.ONE);
+        }
+        for (Map.Entry<String, ItemExpiryPurgerTimeMetricPair> entry : purgerTimeMetricsMap.entrySet()) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **151**. At this line the following field is not protected by synchronization: `purgerTimeMetricsMap`. Your code typically protects this field with synchronization on object: `ItemExpiryPurgerTaskMetricsCollector.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `ItemExpiryPurgerTaskMetricsCollector.this`. For example, like at lines: **101** and **104**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **59**, **101**, and **104**.